enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+  - Pros:
+    - Explicit
+  - Cons:
+    - More complicated schema - weak con
+
+#### How should `microshift pre-run` be executed?
+- `microshift.service` - `ExecStartPre`
+  - No need to add new systemd service files.
+  - It will run on each `systemctl restart microshift` which is not desirable (will it run when systemd restarts MicroShift?)
+- `microshift-pre-run.service`
+  - Running on boot, just once, before `microshift.service`
+  - Not repeated on MicroShift restart
+  - New service file
+
+#### Should MicroShift healthcheck check and log version skew problems so it's easier to debug?
+- Why not?
+
+#### Should backups be kept only for deployments present in ostree command?
+- Is it possible that deployment can be reintroduced?
+  - I.e. it will have the same id and admin might want to rollback to it?
+
+#### Data exists, action is "restore", but backup does not exists
+- It's different from scenario where backup exists, but it fails to be restored
+- See flows below for more context:
+  - ["Rollback to first deployment, failed restore"](#rollback-to-first-deployment-failed-restore)
+  - ["Fail first startup, FDO (FIDO Device Onboard) deployment"](#fail-first-startup-fdo-fido-device-onboard-deployment)
+- Note: this logic is part of `microshift pre-run` which exit code will affect if 
+  `microshift run` executes or not
+- Ideas so far:
+  - Remove the data and allow for `microshift run`
+    - Potential data loss?
+      - MicroShift data is not a biggest problem, it would be more of an inconvenience
+        due to cert regeneration, but static manifests will be applied again
+      - Bigger impact: customer's app loosing reference to PVC it was using
+        - This is also loaded scenario, because if we restore, everything should restore
+          so app shouldn't rely on "future" (from its version perspective) data
+        - Let's assume that thin-snapshot is made, and the same PVC goes across ostree
+          deploys. Rollback/restore would restore thin-snapshot, so app should be
+          just where it was before upgrade, but now it's given a new PVC because TopoLVM
+          relies on in-cluster data which was wiped.
+  - Keep the data and allow for `microshift run`
+    - Scenario: MicroShift attempting to use incompatible (newer) data - **not possible (in theory)**
+      - 1st deployment only ran for 1 boot, so it's up for the 2nd deployment to make its backup
+      - 2nd ostree deployment is staged, booted, etc.
+      - backup fails
+        - redboot -> restore
+        - there's nothing to restore, back to start of the open question
+      - backup succeeds, data is migrated, MicroShift starts, system ends up unhealthy
+        - redboot -> restore
+        - on every following boots will restore (because there's backup) -
+          **(Side question: how do we restore a backup not belonging to THIS deployment?
+            It is now in dir named after PREVIOUS ostree deployment id. Creation datetime?)**
```

Comment:
Is this side question part of resolving this open question? Or is it a separate issue?

Reply:
Separate and currently it's solved with `latest` symlink

Reply:
Why do we need to restore a backup not related to the current deployment?

Reply:
- 2nd deploy boots
- backup of 1st deploy
- red boot
- 2nd deploy boots again
- is told to restore, but what from? There's no deploy2 backup yet

Reply:
Good point. Maybe we need to treat retries of the same deploy differently from the rollback? Instead of setting the mode to "restore" we could set it to "retry".

Reply:
Then we'll need to inspect boot_counter

Reply:
Yes, I think you're right about needing to add more detail. I ignored the retries in the original workflows, but I think we need to include them. Let's not expand _every_ workflow with the retries, though. Let's cover retries as a separate workflow, and include variations like the one here where there is no second backup to restore from. That will keep each scenario simple.

Reply:
Maybe we don't need to check the boot counter explicitly. If we track attempts to boot with the current data, then we can know that we are retrying because the ostree deployment ID will be the same as what is recorded as the last attempt. So when we are told to "restore" and the image is the same and there is no backup we could do nothing.